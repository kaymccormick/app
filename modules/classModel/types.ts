import { Map, List } from 'immutable';
import { ApplicationModule } from '../../src/ApplicationModule';
import {Pojo} from 'classModel';

export interface InputObject {
    [propName: string]: any;
}
export const REQUEST_PROJECT_DATA = 'REQUEST_PROJECT_DATA';
export const LOAD_PROJECT_DATA= 'LOAD_PROJECT_DATA';
export const RECEIVE_PROJECT_DATA = 'RECEIVE_PROJECT_DATA';
export const FETCH_PROJECT_DATA = 'FETCH_PROJECT_DATA';
export const REQUEST_INITIAL_DATA = 'REQUEST_INITIAL_DATA';
export const FETCH_INITIAL_DATA = 'FETCH_INITIAL_DATA';
export const LOAD_INITIAL_DATA = 'LOAD_INITIAL_DATA';
export const RECEIVE_INITIAL_DATA = 'RECEIVE_INITIAL_DATA';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Module extends ApplicationModule<ModuleState> {
}

export interface RequestProjectDataAction {
    type: typeof REQUEST_PROJECT_DATA;
}
export interface RequestInitialDataAction {
    type: typeof REQUEST_INITIAL_DATA;
}
export interface ReceiveInitialDataAction {
    type: typeof RECEIVE_INITIAL_DATA;
    result: InputObject;
}
export interface LoadInitialDataAction {
    type: typeof LOAD_INITIAL_DATA;
}
export interface FetchInitialDataAction {
    type: typeof FETCH_INITIAL_DATA;
}
export interface LoadProjectDataAction {
    type: typeof LOAD_PROJECT_DATA;

}
export interface ReceiveProjectDataAction {
    type: typeof RECEIVE_PROJECT_DATA;
    projects: List<Pojo.ProjectPojo>;
}

export type ActionTypes = ReceiveProjectDataAction | LoadProjectDataAction | RequestProjectDataAction | RequestInitialDataAction |
ReceiveInitialDataAction | LoadInitialDataAction | FetchInitialDataAction;

export type EntitiesType = Map<string, Map<number, Pojo.BasePojo>> ;
export interface ModuleState {
/* Module state fields */
    entities: EntitiesType;
    projects: List<Pojo.ProjectPojo>;
}
