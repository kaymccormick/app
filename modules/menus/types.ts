import { Map, List } from 'immutable';
import { ApplicationModule } from '../../src/ApplicationModule';

export const POPULATE_MENUS = 'POPULATE_MENUS';
export const ADD_MENU_ITEM = 'ADD_MENU_ITEM';
export const SELECT_ITEM = 'SELECT_ITEM';

// eslint-disable-next-line @typescript-eslint/no-empty-interface
export interface Module extends ApplicationModule<ModuleState> {
}

export interface PopulateMenusAction {
  type: typeof POPULATE_MENUS;
}

export interface AddMenuItemAction {
  type: typeof ADD_MENU_ITEM;
  item: MenuItemPojo;
}
export interface SelectItemAction {
  type: typeof SELECT_ITEM;
  item: MenuItemPojo;
}

export type ActionTypes = PopulateMenusAction  | AddMenuItemAction | SelectItemAction /* | other types */

export interface MenuItemPojo {
    parentKey?: string;
    key: string;
    title: string;
    subItems?: Set<string>;
}

export interface ModuleState {
/* Module state fields */
/* After adding a field here, it needs to be returned from
   Module.getInitialState */
    menuItems: Map<string, MenuItemPojo>;
}
