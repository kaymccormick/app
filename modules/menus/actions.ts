import {List} from 'immutable';
import { ActionTypes, SELECT_ITEM, POPULATE_MENUS, ADD_MENU_ITEMS,ADD_MENU_ITEM ,MenuItemPojo } from './types';

export function selectItem(item: MenuItemPojo): ActionTypes {
    return { type: SELECT_ITEM, item };
}

export function addMenuItem(item: MenuItemPojo): ActionTypes {
    return {type: ADD_MENU_ITEM, item};
}

export function addMenuItems(items: List<MenuItemPojo>): ActionTypes {
    return {type: ADD_MENU_ITEMS, items};
}
