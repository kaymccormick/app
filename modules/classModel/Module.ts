import {ApplicationModule, ApplicationModuleArgs} from '../../src/ApplicationModule';
import {Configuration} from '../../src/Configuration';
import {Pojo} from 'classModel';
import reducer from './reducers';
import actions from './actions';
import { WebApplication } from '../../src/WebApplication';
import {ModuleState, EntitiesType} from './types';

import {Map,List,Set} from 'immutable';

import RestClient from '@heptet/rest-client'
/*export type Args = {
  restClient:RestClient,
}*/

export class Module extends ApplicationModule<ModuleState> {
    public restClient?: RestClient;
    public name: string = 'classModel';
    public key: string = 'classModel';

    public constructor(args: ApplicationModuleArgs) {
        super(args);
        this.actions = actions(args.restClient);
        this.restClient = args.restClient;
    }

    public getInitialState(): ModuleState {
        return {
            entities: Map<string, Map<number, Pojo.BasePojo>>(),
        }
    }
    

    public getReducers(): any {
        return reducer({restClient:this.restClient!});

    }
    public setup(app: WebApplication, config: Configuration) {
    }

    public getMainComponent(): Promise<any> {
        return import('./components/MainComponent');
    }
}


