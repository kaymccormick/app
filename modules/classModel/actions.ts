import {List,Map,Set} from 'immutable';
import { Pojo } from 'classModel';
import RestClient from '@heptet/rest-client';
import { EntitiesType, ActionTypes } from './types';
import { addMenuItems } from '../../modules/menus/actions';
import { MenuItemPojo } from '../../modules/menus/types';
import { Actions } from '../../src/types';

import {
    InputObject,
    RECEIVE_INITIAL_DATA,
    REQUEST_INITIAL_DATA,
    RECEIVE_PROJECT_DATA,
    REQUEST_PROJECT_DATA,
    } from './types';

export default (restClient: RestClient): Actions => {
    const receiveInitialData = (result: EntitiesType): ActionTypes =>{
        return { type: RECEIVE_INITIAL_DATA, result};
    }

    const receiveProjectData = (projects: List<Pojo.ProjectPojo>): ActionTypes =>{
        return { type: RECEIVE_PROJECT_DATA, projects };
    }

    const requestInitialData = (): ActionTypes => {
        return { type: REQUEST_INITIAL_DATA };
    }
    const requestProjectData = (): ActionTypes => {
        return { type: REQUEST_PROJECT_DATA };
    }

    const intermediateReceiveInitialData = (result: EntitiesType): (dispatch: any) => void => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (dispatch: any): void => {
            let items = List<MenuItemPojo>();
            const projectsKey = 'projects';
            const projectItem = {title: 'Projects', parentKey: '', key: projectsKey, subItems: Set<string>() };
            result.get('Project').forEach((project: Pojo.ProjectPojo|undefined): void => {
                if(!project) {
                    return;
                }
                const childKey = `${projectsKey}-${project.id}`;
                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                const item = {title: project.name!,
		  parentKey: projectsKey, key: childKey,
		  };
		  projectItem.subItems = projectItem.subItems.add(childKey);
		  items = items.push(item);
            });
            items = items.push(projectItem);
	  dispatch(addMenuItems(items));
            /*            result.get('Module').forEach((module: Pojo.ModulePojo|undefined) => {
                if(!module) {
                    return;
                }
                const parentKey= `${projectsKey}-${module.projectId}`;
                const key = `${parentKey}-${module.id}`;
                const item = {title: module.name!, parentKey, key};
                dispatch(addMenuItem(item));
                dispatch(addMenuItem({title:'Classes', parentKey: key, key: `${key}-classes`}));
                dispatch(addMenuItem({title:'Interfaces', parentKey: key, key: `${key}-interfaces`}));
                dispatch(addMenuItem({title:'Imports', parentKey: key, key: `${key}-imports`}));
                dispatch(addMenuItem({title:'Exports', parentKey: key, key: `${key}-exports`}));
            });
            result.get('Class').forEach((class_: Pojo.ClassPojo|undefined) => {
                if(!class_) {
                    return;
                }
                const module_ = result.get('Module').get(class_.moduleId!) as Pojo.ModulePojo;
                if(!module_) {
                    return;
                }

                const parentKey= `${projectsKey}-${module_.projectId}-${module_.id}-classes`;
                const key = `${parentKey}-${class_.id}`;
                const item = {title: class_.name!, parentKey, key};
                dispatch(addMenuItem(item));
                dispatch(addMenuItem({title:'Methods', parentKey: key, key: `${key}-methods`}));
                dispatch(addMenuItem({title:'Properties', parentKey: key, key: `${key}-props`}));
            });
            
            result.get('Interface').forEach((interface_: Pojo.InterfacePojo|undefined) => {
                if(!interface_) {
                    return;
                }
                const module_ = result.get('Module').get(interface_.moduleId!) as Pojo.ModulePojo;
                if(!module_) {
                    return;
                }

                const parentKey= `${projectsKey}-${module_.projectId}-${module_.id}-interfaces`;
                const key = `${parentKey}-${interface_.id}`;
                const item = {title: interface_.name!, parentKey, key};
                dispatch(addMenuItem(item));
                dispatch(addMenuItem({title:'Methods', parentKey: key, key: `${key}-methods`}));
                dispatch(addMenuItem({title:'Properties', parentKey: key, key: `${key}-props`}));
            });
*/            
            
            dispatch(receiveInitialData(result));
        }
    }


    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const fetchInitialData = (): (dispatch: any) => Promise<any> => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (dispatch: any): Promise<any> => {
            dispatch(requestInitialData());
	    return restClient.getAll().then((result: InputObject): Map<string, Map<number, Pojo.BasePojo>> => {
                let container = Map<string, Map<number, Pojo.BasePojo>>();
                Object.keys(result).forEach((key): void => {
                    const ary = result[key];
		    // how can we sanity check the value?
		    if(ary !== undefined && Array.isArray(ary)) {
		    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
                        let map = Map<number, Pojo.BasePojo>(List<InputObject>(ary).filter((elem): boolean => elem !== undefined).map((elem: InputObject|undefined): [number, Pojo.BasePojo] => [elem!.id, elem!]));
                        container = container.set(key, map);
		    } else{
		    console.log(`invalid value ${ary} for key ${key} in result`)
		    }
		    });
		    return container;
		    }).then((container): void => {
                dispatch(intermediateReceiveInitialData(container));
		    });
        };
    };

    const fetchProjectData = (): (dispatch: any) => void => {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return (dispatch: any): void => {
            dispatch(requestProjectData());
        }
    }

    return {
        requestInitialData,
        intermediateReceiveInitialData,
        fetchInitialData,
        receiveInitialData,
        receiveProjectData,
        fetchProjectData,
    };
};
