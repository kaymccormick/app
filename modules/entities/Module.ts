import uuidv4 from 'uuid/v4';
import {List,Map} from 'immutable';
import {ApplicationModule, ApplicationModuleArgs} from '../../src/ApplicationModule';
import {Configuration} from '../../src/Configuration';
import reducer from './reducers';
import * as actions from './actions';
import RestClient from './RestClient';
/*import EntitiesViewContainer from './components/containers/EntitiesViewContainer';*/
import { WebApplication } from '../../src/WebApplication';
import {EntitiesState,EntityPojo} from './types';

export class Module extends ApplicationModule<EntitiesState> {
    public restClient?: RestClient;
    public name: string = 'entities';
    public key: string = 'entities';
    public id: string;

    public constructor(args: ApplicationModuleArgs) {
        super(args);
        this.id = uuidv4();
        this.actions = actions;
//        console.log(`${this.id} setting rest client`);
        this.restClient = new RestClient({ baseUri: '/cme'});
//        console.log(`${this.id} ${this.restClient}`);
    }

    public getInitialState(): EntitiesState {
        return {
            entities: [],
                entitiesMap: Map<string,EntityPojo>(),
                    ui: { entities: [] },
        }
    }

    public getReducers(): any {
        if(!this.restClient) {
            throw new Error(`${this.id} No rest client`);
        }
        const r = reducer({module: this,restClient:this.restClient});
        console.log(r);
        return r;

    }
    public setup(app: WebApplication, config: Configuration) {
        //const restClient = config.getResource('restClient');
        //this.restClient = restClient;
    }
    public getMainComponent(): Promise<any> {
        return import('./components/MainComponent');
    }
/*    public getMenuTree(): Tree {
    }*/
}

        