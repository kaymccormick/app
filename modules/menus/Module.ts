import uuidv4 from 'uuid/v4';
import {List,Map,Set} from 'immutable';
import {ApplicationModule,ApplicationModuleArgs} from '../../src/ApplicationModule';
import {Configuration} from '../../src/Configuration';
import reducer from './reducers';
import * as actions from './actions';
import { WebApplication } from '../../src/WebApplication';
import {MenuItemPojo, ModuleState} from './types';

export class Module extends ApplicationModule<ModuleState> {
    public name: string = 'menus';
    public key: string = 'menus';
    public id: string;

    public constructor(args: ApplicationModuleArgs) {
        super(args);
        this.id = uuidv4();
        this.actions = actions;
    }

    public getInitialState(): ModuleState {
        return {
            menuItems: Map<string, MenuItemPojo>({'': { key: '', title: 'root',
                subItems: Set<string>(),
            }}),
        }
    }

    public getReducers(): any {
        return reducer({module: this});
    }
    public setup(app: WebApplication, config: Configuration) {
    }
    public getMainComponent(): Promise<any> {
        return import('./components/MainComponent');
    }
}

