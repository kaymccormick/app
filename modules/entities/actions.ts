import{List} from 'immutable';
import { Pojo } from 'classModel';
import RestClient from './RestClient';
import { EntityPojo, REQUEST_ENTITIES, LOAD_ENTITIES, RECEIVE_ENTITIES, ADD_SELECTED_ENTITIES, SELECT_ITEM, EntitiesActionTypes } from './types';
import GetEntitiesResponse from './rest/response/GetEntitiesResponse';
import { addMenuItem } from '../../modules/menus/actions';

export default () => {
const selectItem = (item: any): EntitiesActionTypes => {
    return { type: SELECT_ITEM, item };
};

const addSelectedEntities = (selectedEntities: any): EntitiesActionTypes => {
    return { type: ADD_SELECTED_ENTITIES, selectedEntities };
};

const requestEntities = (): EntitiesActionTypes => {
    return { type: REQUEST_ENTITIES };
}
const receiveEntities = (entities: EntityPojo[]): EntitiesActionTypes => {
    return { type: RECEIVE_ENTITIES, entities };
};
const intermediateReceiveEntities = (entities: EntityPojo[]): (dispatch: any) => void => {
return (dispatch: any) => {
  dispatch(receiveEntities(entities));
  dispatch(addMenuItem({title: 'Entities', parentKey: '', key: 'entities'}));
    entities.forEach((entity) => {
  });
  };
};

const fetchEntities = (restClient: RestClient) => {
    if(!restClient || !restClient.getEntities) {
        console.log(restClient);
        return { type: 'NOOP'};
    }
    return (dispatch: any) => {
        dispatch(requestEntities());
        return restClient.getEntities().then((response: GetEntitiesResponse) => {
            if(response.success) {
                if(response.result) {
                    dispatch(receiveEntities(response.result));
                }
            }
        });
    }

};

const loadEntities = (): EntitiesActionTypes => {
    console.log('load');
    return {type: LOAD_ENTITIES};
};

return { selectItem,
addSelectedEntities,
requestEntities,
receiveEntities,
intermediateReceiveEntities,
fetchEntities,
loadEntities,
};
};
