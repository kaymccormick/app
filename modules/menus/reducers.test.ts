import * as actions from "./actions";
import { Module } from "./Module";
import { AppLogger } from '../../src/AppLogger';
import { MenuItemPojo } from './types';
import makeReducer from "./reducers";
import { List, Set,Map,Seq} from 'immutable';
import RestClient from '@heptet/rest-client'
jest.mock('@heptet/rest-client');

beforeEach((): void => {
// @ts-ignore
    RestClient.mockClear();
});

test("1", () => {
    const restClient = new RestClient({ baseUri: "", logDebug: (arg) => console.log });
    const mod = new Module({ logger: new AppLogger(), restClient });
    const reducer = makeReducer({ module: mod });
    // @ts-ignore
    const state = reducer(undefined, { type: "test" });
    expect(state).toMatchSnapshot();
    let items = List<MenuItemPojo>();
    const testItem = {parentKey: '', title: 'test', key: 'test'};
    items = items.push(testItem);
    const newState = reducer(state, actions.addMenuItems(items));
    expect(newState).toBeDefined();
    const menuItems = newState!.menuItems;
    expect(menuItems).toBeDefined();
    const t = menuItems.get('test');
    expect(t).toBeDefined();
    expect(t).toEqual(testItem);
    let root= menuItems.get('');
    expect(root).toBeDefined();
    root =root!;
    expect(root.subItems!.includes('test')).toBeTruthy();
});
