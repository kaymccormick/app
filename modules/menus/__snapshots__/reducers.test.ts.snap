// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`1 1`] = `
Object {
  "menuItems": Immutable.Map {
    "": Object {
      "key": "",
      "subItems": Immutable.Set [],
      "title": "root",
    },
  },
}
`;
